compilers/concrete-compiler/compiler/lib/Transforms/PreBufferization.cpp

```diff
@@ -0,0 +1,55 @@
+// Part of the Concrete Compiler Project, under the BSD3 License with Zama
```

Comment:
Is that means that tensor.empty is not handled by the bufferization?

Reply:
Correct. The bufferization pass complains about `tensor.empty` and even indicates that the op can be converted to `bufferization.alloc_tensor`. I checked if there is an option to the bufferization pass to apply these conversions automatically, but that does not seem to be the case. Hence the new pre-bufferization pass.

Reply:
Ok indeed 

```
struct EmptyOpInterface
    : public BufferizableOpInterface::ExternalModel<EmptyOpInterface,
                                                    tensor::EmptyOp> {
  bool resultBufferizesToMemoryWrite(Operation *op, OpResult opResult,
                                     const AnalysisState &state) const {
    // The returned tensor does not have specified contents.
    return false;
  }

  LogicalResult bufferize(Operation *op, RewriterBase &rewriter,
                          const BufferizationOptions &options) const {
    // tensor.empty ops are used to indicate the shape of a tensor. They have
    // no defined contents and cannot be bufferized. However, they can be
    // converted to bufferization.alloc_tensor ops, which then bufferize to an
    // allocation (--empty-tensor-to-alloc-tensor).
    return op->emitOpError("cannot be bufferized, but can be converted to "
                           "bufferization.alloc_tensor");
  }
};
```

so could we use directly `bufferizationn.alloc_tensor` instead of `tensor.empty` as we already use it and the pipeline handle it.

Reply:
I would prefer keeping `tensor.empty`.

First, there are already other patterns creating `tensor.empty` ops relatively early in the pipeline. It's just more or less by luck that these are eliminated quite early so that they the pipeline doesn't break.

Second, the semantics are different. While `tensor.empty` just indicates that a new tensor is needed and that initialization doesn't matter, `bufferization.alloc_tensor` always translates to the allocation of a new buffer upon bufferization. So using `bufferization.alloc_tensor` early on effectively prohibits any optimization prior to bufferization that could make use of an existing tensor or that could eliminate the operation. Also, since `tensor.empty` is side-effect-free, it's much more likely that an optimization can eliminate it.

Reply:
That being said, I did not realize that there was already a pass for that. I updated the PR to invoke the pass from the bufferization dialect.